### PR TITLE
CLI: CSF typings for React

### PR DESCRIPTION
**Pull request by tooppaaa**
Issue: #11791

## What I did

Type args

### E2E

1. First remove  `../storybook-e2e-testing/react_typescript-vlatest` if it exists.

2. Then run the e2e test:
```
yarn test:e2e-framework cra_typescript --use-local-sb-cli
```

3. Then link `@storybook/react` in `../storybook-e2e-testing/react_typescript-vlatest` 

4. Then test

```
yarn storybook
yarn build-storybook
yarn start
```

### Official storybook

There's also new stories in `official-storybook` in `typed-button.stories.tsx` that are much easier to play around with for testing autocomplete etc. But doesn't currently test strict type checking.

### CODE REVIEW

**Comment by tooppaaa** [comment]
Left to fix:
```Property 'args' does not exist on type '(args: ButtonProps) => Element'.```

File: lib/addons/src/types.ts [hunk 1]
```diff
@@ -151,3 +151,107 @@ export interface ClientStoryApi<StoryFnReturnType = unknown> {
...
+  args?: Partial<Args>;
+
+  /**
+   * ArgTypes encode basic metadata for args, such as `name`, `description`, `defaultValue` for an arg. These get automatically filled in by Storybook Docs.
+   * @see [Control annotations](https://github.com/storybookjs/storybook/blob/91e9dee33faa8eff0b342a366845de7100415367/addons/controls/README.md#control-annotations)
+   */
+  argTypes?: ArgTypes;
+
+  /**
+   * Wrapper components or Storybook decorators that wrap a story.
```
Comment (yannbf) [inline lib/addons/src/types.ts:175]:
```suggestion
   * Custom metadata for a story.
```

File: examples/official-storybook/stories/demo/typed-button.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import { Meta, Story } from '@storybook/react/types-6-0';
+import TsButton from '../../components/TsButton';
+
+export default {
+  title: 'Other/Demo/TsButton',
+  component: TsButton,
+  decorators: [
+    (StoryFn) => (
+      <>
+        <StoryFn />
```
Comment (yannbf) [inline examples/official-storybook/stories/demo/typed-button.stories.tsx:11]: Would this also work with `{StoryFn()}` ? I don't know exactly why two different ways could be used but I assume both should work with no type issues?

Comment (shilman) [reply examples/official-storybook/stories/demo/typed-button.stories.tsx:11]: Yep tested

**Review by shilman — APPROVED** [review]
LGTM. Merging! Expect we'll need to iterate.... 🤞

**Pull request merged**